compiler/aarch64/codegen/OMRTreeEvaluator.cpp

```diff
@@ -1268,31 +1268,460 @@ OMR::ARM64::TreeEvaluator::mstoreiEvaluator(TR::Node *node, TR::CodeGenerator *c
+   generateTrg1Src1Instruction(cg, addvOp, node, tempReg, tempReg);
+   generateMovVectorElementToGPRInstruction(cg, TR::InstOpCode::umovwb, node, resReg, tempReg, 0);
+
+   node->setRegister(resReg);
+   cg->stopUsingRegister(tempReg);
+   cg->decReferenceCount(firstChild);
+
+   return resReg;
    }
 
 TR::Register*
 OMR::ARM64::TreeEvaluator::mFirstTrueEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::unImpOpEvaluator(node, cg);
+   TR::Node *firstChild = node->getFirstChild();
+   TR_ASSERT_FATAL_WITH_NODE(node, firstChild->getDataType().getVectorLength() == TR::VectorLength128,
+                   "Only 128-bit vectors are supported %s", firstChild->getDataType().toString());
+
+   TR::DataType et = firstChild->getDataType().getVectorElementType();
+   TR::Register *maskReg = cg->evaluate(firstChild);
+   TR::Register *temp1Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *resReg = cg->allocateRegister();
+
+   /*
+    * This evaluator returns the index of the first lane that is set.
+    * If all lanes are not set, then the number of the lanes is returned.
+    */
+   switch (et)
+      {
+      case TR::Int8:
+         /*
+          * shrn    v1.8b, v0.8h, #7            ; Moves mask values to bit 0 and 1 of the lane 0 - 7.
+          * sli     v1.16b, v1.16b, #6          ; Shifts left v1 by 6 bits and inserts into v1. bit 6 - 9 of the lane 0, 1, 2 and 3 have mask values.
+          * ushr    v1.8h, v1.8h, #6            ; Moves mask values to bit 0 - 3 of each lane.
+          * sli     v1.8h, v1.8h, #12           ; Shifts left v1 by 12 bits and inserts into v1. bit 12 - 19 of the lane 0 and 1 have mask values.
+          * umov    x0, v1.2d[0]                ; Mask values at bit 12 - 19 and bit 44 - 51.
+          * bfi     x0, x0, #24, #20            ; Inserts bit 0 - 19 into bit 24 - 43.
+          * ubfx    x0, x0, #36, #16            ; Moves bit 36 - 51 to bit 0 - 15.
+          * orr     w0, w0, #0x10000            ; Sets the bit 16.
+          * rbit    w0, w0                      ; Reverses bits.
+          * clz     w0, w0                      ; Counts leading zeros.
+          */
+         generateVectorShiftImmediateInstruction(cg, TR::InstOpCode::vshrn_8b, node, temp1Reg, maskReg, 7);
+         generateVectorShiftImmediateInstruction(cg, TR::InstOpCode::vsli16b, node, temp1Reg, temp1Reg, 6);
+         generateVectorShiftImmediateInstruction(cg, TR::InstOpCode::vushr8h, node, temp1Reg, temp1Reg, 6);
+         generateVectorShiftImmediateInstruction(cg, TR::InstOpCode::vsli8h, node, temp1Reg, temp1Reg, 12);
+         generateMovVectorElementToGPRInstruction(cg, TR::InstOpCode::umovxd, node, resReg, temp1Reg, 0);
+         generateBFIInstruction(cg, node, resReg, resReg, 24, 20, true);
+         generateUBFXInstruction(cg, node, resReg, resReg, 36, 16, true);
+         generateLogicalImmInstruction(cg, TR::InstOpCode::orrimmw, node, resReg, resReg, false, 0x400);
```

Comment:
Please add a comment for `0x400` and the other values for `generateLogicalImmInstruction()` below.

Reply:
Added comments at the lines with `generateLocalImmInstruction()` call.